### PR TITLE
Test

### PR DESCRIPTION
**Pull request by deathbearbrown**
### Description of the problem this pull request fixes

Adding Karma, Mocha, Chai, Sinon, Enzyme stack for testing and jsx template support.

Right now this just has a hello world script and the other scripts are commented out.  I started fixing the other scripts, but realized that the model is running before everything runs, and I know we are planning to use redux (or something) and change the model soon.  I'd rather wait until that happens to fix the scripts.

To test:
- npm install
- npm test

Also test for regressions:
- npm run build
- npm start

### CODE REVIEW

File: karma.config.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    // include some polyfills for babel and phantomjs
+    files: [
+      'node_modules/babel-polyfill/dist/polyfill.js',
+      './node_modules/phantomjs-polyfill/bind-polyfill.js',
+      './test/**/*.js' // specify files to watch for tests
+    ],
+    preprocessors: {
+      // these files we want to be precompiled with webpack
+      // also run tests throug sourcemap for easier debugging
```
Comment (kadamwhite) [inline karma.config.js]: 'through'

File: karma.config.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      ['./test/**/*.js']: ['webpack', 'sourcemap']
+    },
+    webpack: {
+       devtool: 'inline-source-map',
```
Comment (kadamwhite) [inline karma.config.js:25]: After reading up more on devtools for #271 I want to do a speed test b/w this and `cheap-module-inline-source-map`

File: package.json [hunk 3]
```diff
@@ -23,27 +23,44 @@
...
+    "babel-preset-react": "^6.3.13",
+    "babel-register": "^6.3.13",
     "chai": "^3.0.0",
-    "chai-spies": "^0.7.1",
     "d3": "^3.5.6",
     "d3-geo-projection": "^0.2.15",
     "datalib": "^1.4.6",
     "deep-diff": "^0.3.3",
+    "enzyme": "^1.2.0",
+    "es6-promise": "^3.0.2",
```
Comment (kadamwhite) [inline package.json]: This doesn't look like it's used, may be extraneous

File: package.json [hunk 4]
```diff
@@ -23,27 +23,44 @@
...
     "es6-promisify": "^3.0.0",
     "eslint": "^1.10.3",
     "eslint-loader": "^1.2.1",
     "eslint-plugin-react": "^3.16.1",
+    "fs": "0.0.2",
     "istanbul": "^0.3.21",
     "jshint": "^2.8.0",
+    "json-loader": "^0.5.4",
+    "jstransform": "^11.0.3",
```
Comment (kadamwhite) [inline package.json]: This is another one that was removed in #266; it's unclear whether it was added back in to support this build or due to a merge with old code

File: test/test_helper.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
+// no longer being used once Karma is set up
+var expect = require('chai').expect,
+    sinon = require('sinon'),
+    fs = require('fs'),
+    res = './test/resources/';
+
+global.expect = expect;
+global.sinon = sinon;
```
Comment (kadamwhite) [inline test/test_helper.js]: Do these need to be global?

**Comment by kadamwhite** [comment]
@arvind It looks like on other Vega projects the team switched from `filename.spec.js` to `filename.test.js` -- This goes the other direction, do you have a strong preference either way?

**Comment by arvind** [comment]
`filename.test.js` would be my preference for consistency.

**Comment by kadamwhite** [comment]
@deathbearbrown :+1: from me on regression & testing, this is working -- thanks so much for taking on setting this up, I know these tools took a while to configure! I have some naive questions in the comments above, and Arvind's preference is to maintain `.test.js` as the extension convention, but this should be able to :ship: with those changes.

File: package.json [hunk 6]
```diff
@@ -58,7 +75,8 @@
     "watch": "npm run watch:js & npm run watch:scss",
     "watch:js": "webpack -d --watch",
     "watch:scss": "nodemon -e src/scss -x \"npm run build:scss\"",
-    "test:js": "mocha --timeout 30000 --recursive test/",
+    "test": "node_modules/.bin/karma start karma.config.js",
```
Comment (arvind) [inline package.json]: `npm` automatically puts `node_modules/.bin` in the path for `npm run` scripts. So you should be able to define this as `karma start karma.config.js` directly.

File: package.json (hunk 4 above)
Comment (deathbearbrown) [inline package.json]: json-loader should be there I think jstransform was a dependency of babel-es2015 which I removed

File: package.json (hunk 3 above)
Comment (deathbearbrown) [inline package.json]: yeah, that's a dependency of something I removed >.<

**Comment by deathbearbrown** [comment]
updated to file.test.js

File: karma.config.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    // include some polyfills for babel and phantomjs
+    files: [
+      'node_modules/babel-polyfill/dist/polyfill.js',
+      './node_modules/phantomjs-polyfill/bind-polyfill.js',
+      './test/**/*.js' // specify files to watch for tests
+    ],
+    preprocessors: {
+      // these files we want to be precompiled with webpack
+      // also run tests through sourcemap for easier debugging
+      ['./test/**/*.js']: ['webpack', 'sourcemap']
```
Comment (arvind) [inline karma.config.js]: I get an error thrown on this line: 

```
03 03 2016 14:53:37.940:ERROR [config]: Invalid config file!
  /Users/arvind/Research/lyra/code/karma.config.js:22
      ['./test/**/*.js']: ['webpack', 'sourcemap']
      ^
SyntaxError: Unexpected token [
```

Should this be valid JSON (`'./test/**/*.js'`)?

File: karma.config.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
+var argv = require('yargs').argv;
```
Comment (arvind) [inline karma.config.js:1]: `yargs` is also not defined as a `devDependency` in `package.json`.

File: karma.config.js (hunk 7 above)
Comment (deathbearbrown) [inline karma.config.js]: hrmm... lemme see if I can recreate that.  Is this after running `npm test` ?

**Comment by arvind** [comment]
If I resolve my two inline comments and then `npm run test`, I see 

```
PhantomJS 1.9.8 (Mac OS X 0.0.0) LOG LOG: 'I'm `fs` modules'

  hello world
    ✓ works!

PhantomJS 1.9.8 (Mac OS X 0.0.0): Executed 1 of 1 SUCCESS (0.003 secs / 0 secs)
TOTAL: 1 SUCCESS
```

Except, in the prior output, there is 

```
ERROR in ./src/js/components/index.jsx
Module build failed: SyntaxError: /Users/arvind/Research/lyra/code/src/js/components/index.jsx: Unexpected token (7:2)
   5 |
   6 | module.exports = ReactDOM.render(
>  7 |   <Sidebars />,
     |   ^
   8 |   d3.select('#sidebars').node()
   9 | );
  10 |
    at Parser.pp.raise (/Users/arvind/Research/lyra/code/node_modules/babel-core/node_modules/babylon/index.js:1412:13)
...
```

Is it problematic for the test runner to return a successful result if there was an intermediate build failure even if it's not testing that file specifically?

File: karma.config.js (hunk 8 above)
Comment (deathbearbrown) [inline karma.config.js:1]: it's a dependency of webpack.

Comment (arvind) [inline karma.config.js:1]: Hmm, I'm on node `0.12.10` and npm `2.14.9` and on clean `npm install` of this branch a subsequent `npm run test` throws this error (assuming the above config error is fixed). 

```
03 03 2016 15:19:39.578:ERROR [config]: Invalid config file!
  Error: Cannot find module 'yargs'
```

**Comment by deathbearbrown** [comment]
That's my bad!!! I didn't copy over the presets in the karma.config!!

**Comment by deathbearbrown** [comment]
You should not see that syntax error any more

**Comment by deathbearbrown** [comment]
I added a little test for primative/Guide.js to make sure everything is cool. I will work on adding new tests every morning until we are caught up. :)

**Pull request merged**